### PR TITLE
Jax-jit can use device vjp

### PR DESCRIPTION
**Pull request by albi3ro**
[sc-50418]

This PR binds device-provided VJPs to the JAX-JIT interface. It also switches the jax-jit interface to the "Jacobian Product Calculator" system of defining how to take Jacobians/ JVPs/ VJPs.

Since we need to place a pure-callback around the jacobian calculation, but we want to calculate results and jacobians at the same time when feasible, this PR adds the `JacobianProductCalculator.execute_and_compute_jacobian` method. This has the benefit over calling `execute_fn` and `compute_jacobian` separately, as it is always `grad_on_execution=True` when actually taking derivatives, and always `grad_on_execution=False` when purely calculating results.

### CODE REVIEW

File: pennylane/interfaces/jax_jit.py [hunk 1]
```diff
@@ -91,174 +105,114 @@ def _jac_shape_dtype_struct(tape: "qml.tape.QuantumScript", device: "qml.Device"
...
+        # the shape without the vmap dimensions, while the function here will be
+        # executed with objects containing the vmap dimensions. So res[i].ndim
+        # will have an extra dimension for every `jax.vmap` wrapping this execution.
+        #
+        # The execute_fn will return an object with shape `(n_observables, batches)`
+        # but the default behaviour for `jax.pure_callback` is to add the extra
+        # dimension at the beginning, so `(batches, n_observables)`. So in here
+        # we detect with the heuristic above if we are executing under vmap and we
+        # swap the order in that case.
+        return jax.tree_map(lambda r, s: r.T if r.ndim > s.ndim else r, res, shape_dtype_structs)
```
Comment (dwierichs) [inline pennylane/interfaces/jax_jit.py:142]: Good catch :sweat_smile: 
Just to make sure: Does this transpose do the right reordering for multiple `vmap`s as well? :)

File: pennylane/interfaces/jax_jit.py [hunk 2]
```diff
@@ -91,174 +105,114 @@ def _jac_shape_dtype_struct(tape: "qml.tape.QuantumScript", device: "qml.Device"
...
-        jacs = res_processing_fn(jacs)
-        jacobians.append(jacs)
-    return tuple(jacobians)
+_execute_vjp_jit = jax.custom_vjp(_execute_wrapper, nondiff_argnums=[1, 2, 3, 4])
+_execute_vjp_jit.defvjp(_vjp_fwd, _vjp_bwd)
 
 
-def execute(tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n=1, max_diff=1):
-    """Execute a batch of tapes with JAX parameters on a device.
+def jax_jvp_jit_execute(tapes, execute_fn, jpc, device):
```
Comment (dwierichs) [inline pennylane/interfaces/jax_jit.py:201]: I would have expected the name to be `jax_jit_jvp_execute`, because both `jax` and `jit` are part of the interface name.

File: tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py [hunk 3]
```diff
@@ -46,19 +51,28 @@
 
 
 @pytest.mark.parametrize(
-    "interface,dev,diff_method,grad_on_execution", interface_and_qubit_device_and_diff_method
+    "interface,dev,diff_method,grad_on_execution, device_vjp",
```
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py:54]:
```suggestion
    "interface,dev,diff_method,grad_on_execution,device_vjp",
```

File: tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py [hunk 4]
```diff
@@ -46,19 +51,28 @@
...
     """Test that using the QNode with JAX integrates with the PennyLane
     stack"""
 
-    def test_execution_with_interface(self, dev, diff_method, grad_on_execution, interface):
+    def test_execution_with_interface(
+        self, dev, diff_method, grad_on_execution, interface, device_vjp
+    ):
         """Test execution works with the interface"""
         if diff_method == "backprop":
             pytest.skip("Test does not support backprop")
```
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py:66]: I know this was here before, but why does this test not support backprop? :thinking:

Comment (dwierichs) [reply tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py:66]: Just ran the test, and it passes :upside_down_face: (same for non-JIT)

File: tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py [hunk 5]
```diff
@@ -956,16 +1054,18 @@ def expected_circuit(theta):
...
-    def test_second_derivative(self, dev, diff_method, grad_on_execution, interface, tol):
+    def test_second_derivative(
+        self, dev, diff_method, grad_on_execution, device_vjp, interface, tol
+    ):
         """Test second derivative calculation of a scalar-valued QNode"""
-
         gradient_kwargs = {}
-        if diff_method == "adjoint":
+        if diff_method in {"adjoint", "device"}:
             pytest.skip("Adjoint does not second derivative.")
```
Comment (dwierichs) [inline tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py:1069]:
```suggestion
            pytest.skip("Adjoint does not support second derivatives.")
```

File: tests/interfaces/test_jacobian_products.py [hunk 6]
```diff
@@ -349,6 +361,66 @@ def test_batch_jacobian(self, jpc, shots):
                 jacs[1][1][1], -np.sin(x) * np.cos(y), atol=_tol_for_shots(shots)
             )
 
+    def test_batch_execute_jacobian(self, jpc, shots):
+        """Test compute_jacobian on a batch with ragged observables and parameters."""
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:365]:
```suggestion
        """Test execute_and_compute_jacobian on a batch with ragged observables and parameters."""
```

File: tests/interfaces/test_jacobian_products.py [hunk 7]
```diff
@@ -349,6 +361,66 @@ def test_batch_jacobian(self, jpc, shots):
...
+
+                assert qml.math.allclose(jacs[0][i], -np.sin(phi), atol=_tol_for_shots(shots))
+                assert qml.math.allclose(jacs[1][i][0][0], 0, atol=_tol_for_shots(shots))
+                assert qml.math.allclose(jacs[1][i][0][1], np.cos(y), atol=_tol_for_shots(shots))
+                assert qml.math.allclose(
+                    jacs[1][i][1][0], -np.cos(x) * np.sin(y), atol=_tol_for_shots(shots)
+                )
+                assert qml.math.allclose(
+                    jacs[1][i][1][1], -np.sin(x) * np.cos(y), atol=_tol_for_shots(shots)
+                )
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:408]: For consistency
```suggestion
                assert qml.math.allclose(jacs[1][i][0][0], 0, atol=_tol_for_shots(shots))
                assert qml.math.allclose(jacs[1][i][0][1], np.cos(y), atol=_tol_for_shots(shots))
                assert qml.math.allclose(
                    jacs[1][i][1][0], -np.cos(x) * np.sin(y), atol=_tol_for_shots(shots)
                )
                assert qml.math.allclose(
                    jacs[1][i][1][1], -np.sin(x) * np.cos(y), atol=_tol_for_shots(shots)
                )
                assert qml.math.allclose(jacs[0][i], -np.sin(phi), atol=_tol_for_shots(shots))
```

File: tests/interfaces/test_jacobian_products.py [hunk 8]
```diff
@@ -520,6 +617,50 @@ def test_execute_jvp_multi_params_multi_out(self, jpc):
         assert qml.math.allclose(jvp[1][0], 0)
         assert qml.math.allclose(jvp[1][1], -0.6 * np.sin(phi))
 
+    def test_execute_jacobian_multi_params_multi_out(self, jpc):
+        """Test execute_and_compute_jvp with multiple parameters and multiple outputs"""
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:621]:
```suggestion
        """Test execute_and_compute_jacobian with multiple parameters and multiple outputs"""
```

File: tests/interfaces/test_jax_jit.py [hunk 9]
```diff
@@ -137,13 +137,15 @@ def cost(a):
 
         # adjoint method only performs a single device execution, but gets both result and gradient
```
Comment (dwierichs) [inline tests/interfaces/test_jax_jit.py:138]: Does this comment still apply? :thinking:

**Review by dwierichs — COMMENTED** [review]
Only had a few minor comments, otherwise looks good to me! 
Again, amazing to see the cleaned up logic! :100:

Those test changes though :sweat_smile: 

Do we want a changelog entry remarking that gradients are not precalculated if not requested (c.f. changed test in `test_jax_jit.py`)?

File: pennylane/interfaces/jax_jit.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/interfaces/jax_jit.py:142]: Copy and paste from the previous version.  I don't technically understand the change.

File: pennylane/interfaces/jacobian_products.py [hunk 10]
```diff
@@ -179,6 +179,33 @@ def compute_jacobian(self, tapes: Batch) -> Tuple:
...
+        >>> tape1 = qml.tape.QuantumScript([qml.RY(0.2, wires=0)], [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliX(0))])
+        >>> batch = (tape0, tape1)
+        >>> results, jacs = jpc.execute_and_compute_jacobian(batch)
+        >>> results
+        (0.9950041652780258, (0.9800665778412417, 0.19866933079506116))
+        >>> jacs
+        (array(-0.09983342), (array(-0.19866933), array(0.98006658)))
+
+        While this method could support non-scalar parameters in theory, no implementation currently supports
+        jacobians with non-scalar parameters.
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:205]: It starts with `this method`, and the method accepts tapes?

File: pennylane/interfaces/torch.py [hunk 11]
```diff
@@ -189,7 +189,8 @@ def backward(ctx, *dy):
         return (None,) + vjps
 
 
-def execute(tapes, execute_fn, jpc):
+# pylint: disable=unused-argument
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/torch.py:192]: pylint was being extremely finicky with not picking up on this for some reason, so I'm not going to touch it again.

File: tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py (hunk 4 above)
Comment (albi3ro) [reply tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py:66]: I have no idea?

**Comment by albi3ro** [comment]
Ran some performance numbers 🚀 

Screenshot 2023-12-14 at 9 50 52 AM

Shows similar behaviour to the timings for torch and autograd. 

Interestingly enough, we have support for jacobians with jit, even though we don't with non-jit.  I think this is because we aren't setting `vectorized=True` in the pure-callback around the vjp calculation.  Something to revisit once the device derivative supports a broadcast dimension.

File: tests/interfaces/test_jax_jit.py (hunk 9 above)
Comment (albi3ro) [reply tests/interfaces/test_jax_jit.py:138]: updated.

File: tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py [hunk 12]
```diff
@@ -46,19 +51,26 @@
...
 
-    def test_execution_with_interface(self, dev, diff_method, grad_on_execution, interface):
+    def test_execution_with_interface(
+        self, dev, diff_method, grad_on_execution, interface, device_vjp
+    ):
         """Test execution works with the interface"""
-        if diff_method == "backprop":
-            pytest.skip("Test does not support backprop")
+        if dev.name == "param_shift.qubit":
+            pytest.xfail("gradient transforms have a different vjp shape convention")
```
Comment (timmysilv) [inline tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py:66]: is this being tracked elsewhere?

**Review by timmysilv — COMMENTED** [review]
I'm really bothered by this addition of a device kwarg for some reason... because we have DeviceDerivatives but it's something else? idk. anyway this is really great, will be so close after this!

File: pennylane/interfaces/autograd.py [hunk 13]
```diff
@@ -97,10 +97,12 @@ def grad_fn(dy):
 logger.addHandler(logging.NullHandler())
 
 
+# pylint: disable=unused-argument
 def autograd_execute(
     tapes: Batch,
     execute_fn: ExecuteFn,
     jpc: qml.interfaces.jacobian_products.JacobianProductCalculator,
+    device=None,
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/autograd.py:105]: I do think it could be feasible to remove this in the future if we either:

A) No longer need `device` to determine the shape
B) Directly wrap `execute_fn` and `JacobianProductCalculator` before even sending it to the interface code.

But this is a step in the correct direction and gets the required features.

File: tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py (hunk 12 above)
Comment (albi3ro) [reply tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py:66]: I have a technical story for v0.35 for "VJP's have a well defined, consistent shape".

File: pennylane/interfaces/jacobian_products.py [hunk 14]
```diff
@@ -584,6 +606,33 @@ def compute_jacobian(self, tapes):
...
+    def execute_and_compute_jacobian(self, tapes):
+        if tapes not in self._results_cache and tapes not in self._jacs_cache:
+            results, jacs = self._dev_execute_and_compute_derivatives(tapes)
+            self._results_cache[tapes] = results
+            self._jacs_cache[tapes] = jacs
+            return results, jacs
+
+        if tapes not in self._jacs_cache:
+            # Here the jac was not cached but the results were. This can not happen because results are never
+            # cached alone (note that in the else clause above computing only results, jac must already be present)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:618]: Ah. this is a byproduct of me moving the order the branches.
```suggestion
            # cached alone (note that in the else clause below computing only results, jac must already be present)
```

File: pennylane/interfaces/autograd.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/autograd.py:105]: And I'll probably have to add an `is_differentiable` keyword argument for tensorflow 😢 though I'm still trying to figure out if there's a better way.

File: pennylane/interfaces/execution.py [hunk 15]
```diff
@@ -758,13 +761,6 @@ def device_gradient_fn(inner_tapes, **gradient_kwargs):
 
             gradient_fn = device_gradient_fn
 
-            # Adjoint Jacobian with backward pass and jitting needs the original circuit output state which
-            # can not be reused from the device if `grad_on_execution is False`.
-            if interface == "jax-jit":
-                use_device_state = gradient_kwargs.get("use_device_state", None)
-                if use_device_state:
-                    gradient_kwargs["use_device_state"] = False
```
Comment (timmysilv) [inline pennylane/interfaces/execution.py:766]: is this logic no longer used? seems like it's relevant for `QubitDevice.adjoint_jacobian`

**Review by timmysilv — APPROVED** [review]
one more questio, but assuming it isn't a problem, sounds good!

File: pennylane/interfaces/execution.py (hunk 15 above)
Comment (albi3ro) [reply pennylane/interfaces/execution.py:766]: We are now in a situation where we either just calculate the results `cost(x)` or we always calculate the results and gradients at the same time `jax.jacobian(cost)(x)`. We no longer have the case where we do the forward pass, leave the pure callback, and then come back and calculate the jacobian later. So we could technically just set `use_device_state=True` for all cases.

Comment (josh146) [reply pennylane/interfaces/execution.py:766]: @albi3ro since this this looks like interface-agnostic code, what about other interfaces, like PyTorch? E.g.,

```python
res = cost(x)
... # more computation
# later
res.backward()
```

Here, unless something else has changed, we have a case where the forward pass occurs, and then later the gradients are requested.

Comment (albi3ro) [reply pennylane/interfaces/execution.py:766]: This was a fix specific to JAX (see removed line 763).
